Tutorial about developing with cmake and colcon (#145)
ignitionrobotics/ign-cmake: tutorials/developing_with_ign-cmake.md

\page developingwithcmake Developing with Ignition CMake 

# Developing with Ignition CMake

This tutorial documents various tips and strategies for developing with Ignition CMake.

## Helpful CMake flags

There are several flags that control the results of the CMake tool.
Some of these flags are built into CMake, where some are Ignition CMake specific.

### Setting the build type

The `CMAKE_BUILD_TYPE` variable controls the type of binary output from the build stage.
This will have an impact on the flags passed to the compiler and linker.

The available options are:

  * `RelWithDebInfo` (default): Mostly optimized build, but with debug symbols.
  * `Debug`: Debug build without optimizations, with debug symbols enabled.
  * `Release`: Fully optimized build, no debug symbols.
  * `MinSizeRel`: Fully optimized build, minimal binary size 
  * `Coverage`: Build with additional information required for code coverage computation

>   * `Profile`: Use flags that are helpful with the `gprof` profiling tool

Reviewer: Can we add an example of how to set one of these flags, like you did with `CMAKE_EXPORT_COMPILE_COMMANDS` in the [Creating a compilation database](https://github.com/ignitionrobotics/ign-cmake/blob/217011749bfb00a56bd3d46e0a7d1c9aedf47eeb/tutorials/developing_with_ign-cmake.md#creating-a-compilation-database) section? I'd think that including something like `-DCMAKE_BUILD_TYPE=Debug` would be sufficient.

>   * `Profile`: Use flags that are helpful with the `gprof` profiling tool

Author: I just went with wikipedia, since it's more likely to be constant.

More information about flags applied can be found in [IgnSetCompilerFlags.cmake](https://github.com/ignitionrobotics/ign-cmake/blob/ign-cmake2/cmake/IgnSetCompilerFlags.cmake)

### Creating a compilation database

> `CMake` can optionally generate a compilation data base that may be used with a variety of code completion tools.

Reviewer: ```suggestion
`CMake` can optionally generate a compilation database that may be used with a variety of code completion tools.
```

Should we include a link to compilation databases as well? I've always referred to this one, but maybe there's something better out there: https://clang.llvm.org/docs/JSONCompilationDatabase.html

To enable this set: `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON`.

### Using the ninja build system

Rather than using `make`, it may be desired to use the [Ninja](https://ninja-build.org/) build tool.

This may be applied by setting `-GNinja`.

### Address sanitizer

The `gcc` and `clang` compilers have a set of flags to generate instrumented builds for detecting memory leaks.

```
-DCMAKE_CXX_FLAGS:STRING= -fsanitize=address  -fsanitize=leak -g
-DCMAKE_C_FLAGS:STRING=-fsanitize=address  -fsanitize=leak -g
-DCMAKE_EXE_LINKER_FLAGS:STRING=-fsanitize=address  -fsanitize=leak

> -DCMAKE_MODULE_LINKER_FLAGS:STRING=-fsanitize=address  -fsanitize=leak

Reviewer: A few questions came to my mind when reading through this. Perhaps we can address them in the tutorial:
1. What are the differences between all of these flags?
2. Should users use all of these flags to check for memory leaks, or can they pick and choose particular flags based on their needs?

Author: All of the flags are required for address and leak checking.  I updated the wording to indicate that.

```

This will report if memory is leaked during execution of binaries or tests.

### Using CCache

When you are doing frequent rebuilds, you can use a program to cache intermediate compiler results.

> First, install `ccache` and configure it to an appropriate cache size for your system:

Reviewer: ```suggestion
First, install [`ccache`](https://ccache.dev/) and configure it to an appropriate cache size for your system:
```

Author: fb94592

```

> $ sudo apt install ccache

Reviewer: ```suggestion
$ sudo apt update && sudo apt install ccache
```

Author: fb94592

$ ccache -M10G
Set cache size limit to 10.0 GB
```

> Then set the CMake flag either via the command line or `defaults.yaml`:

Reviewer 2: This file is only explained below, so how about pointing to the default path here just to give a hint that this is a colcon thing?

```suggestion
Then set the CMake flag either via the command line or `~/.colcon/defaults.yaml`:
```

Author: I reworded the intro section to explain that the colcon section is coming later and dropped defaults from here.

> Then set the CMake flag either via the command line or `defaults.yaml`:

Reviewer: Users might not know what `defaults.yaml` is yet, since the `using a defaults file` section appears later in this tutorial. Perhaps we can add a note to this line that refers to the `using a defaults file` section?

Author: Changed the wording since @chapulina had the same concern.

```
"-DCMAKE_C_COMPILER_LAUNCHER=ccache",
"-DCMAKE_CXX_COMPILER_LAUNCHER=ccache"
```

> ### Enabling/Disabling Documentation

Reviewer 2: You could also have a section for `BUILD_TESTING`, and `BUILDSYSTEM_TESTING`

When you are doing frequent rebuilds, it generally doesn't make sense to rebuild documentation each build.

> To disable building documentation (default is on), set the CMake flag either via the command line or `defaults.yaml`:

Reviewer 2: ```suggestion
To disable building documentation (default is on), set the CMake flag either via the command line or `~/.colcon/defaults.yaml`:
```

Author: Same here.

> To disable building documentation (default is on), set the CMake flag either via the command line or `defaults.yaml`:

Reviewer: Same comment here about `defaults.yaml` as I mentioned above.

```
"-DBUILD_DOCS=OFF"
```

## Developing with Colcon and vcstool

[`colcon`](https://colcon.readthedocs.io/en/released/) is a tool that improves the workflow of building and testing multiple software packages.
As an Ignition collection is composed of multiple packages that are frequently built and tested together, `colcon` eases this workflow.

The basic outline of obtaining Ignition source packages via `vcs` and building with `colcon` is available in the [Ignition source installation documentation](https://ignitionrobotics.org/docs/latest/install_ubuntu_src).

### Passing CMake flags via command line

> When performing `colcon` builds, flags may be passed to Ignition CMake to configure the build 

Reviewer: ```suggestion
When performing `colcon` builds, flags may be passed to Ignition CMake to configure the build.
```

Author: fb94592

This can be done via the `--cmake-args` flag in `colcon`:

```
colcon build --cmake-args -DCMAKE_EXPORT_COMPILE_COMMANDS=1
```

### Passing colcon mixins via command line

To ease configuration of common flags, `colcon` has a concept of `mixins`, that are flags that "shortcut" groups of behavior.

The set of readily-available defaults is in the [colcon-mixin-repository](https://github.com/colcon/colcon-mixin-repository).

To install:

``` 
$ colcon mixin add default https://raw.githubusercontent.com/colcon/colcon-mixin-repository/master/index.yaml
$ colcon mixin update default
$ colcon mixin show
```

To use:

```

> colcon build --mixin ccache rel-with-deb-info

Reviewer: I think it would be useful to mention what `ccache` and `rel-with-deb-info` do here. For example, I am assuming that `ccache` runs the cmake build args listed in [ccache.mixin](https://github.com/colcon/colcon-mixin-repository/blob/master/ccache.mixin), right?

Author: fb94592

```
### Using a defaults file

It is useful to be able to apply a consistent set of flags across an entire Ignition collection when building.
One mechanism for accomplishing this is a `defaults.yaml` file.
This is a file of configuration options that `colcon` will read to customize the default behavior.
More information about the `defaults.yaml` file can be found in the corresponding [`colcon` documentation](https://colcon.readthedocs.io/en/released/user/configuration.html#defaults-yaml)

To try this out, first create an Ignition source workspace:
```
mkdir -p ~/ign_edifice/src
cd ~/ign_edifice/
wget https://raw.githubusercontent.com/ignition-tooling/gazebodistro/master/collection-edifice.yaml

> vcs import src < collection-edifice.yaml

Reviewer: We should add a note about installing `vcs` in case users don't have this installed already.

Author: I have a note in the level 2 section that redirects to the ignition source installation documentation that should cover this, is that sufficient?

Reviewer: I had the same thought after I commented. Perhaps we can just mention users should go back to section 2 if they don't have `vcs` installed yet?

```

Then add a `~/ign_edifice/defaults.yaml` file with compilation flags:

```
{
  "build": {
    "merge-install": true,
    "symlink-install": true,
    "cmake-args": [
      "--no-warn-unused-cli",
      "-DCMAKE_EXPORT_COMPILE_COMMANDS=1",
      "-DCMAKE_BUILD_TYPE=RelWithDebInfo",
    ]
  }
}
```

To build with this defaults file, first export the correct environment variable and execute colcon:

```
cd ~/ign_edifice
export COLCON_DEFAULTS_FILE=`pwd`/defaults.yaml
colcon build
```

### Using a defaults file with mixins

Mixins can also be applied via the `defaults.yaml` file:

```
{
  "build": {
    "merge-install": true,
    "symlink-install": true,
    "mixins": [
      "compile-commands",
      "rel-with-deb-info",
    ]
  }
}
```

### Setting a per-workspace defaults file with direnv

Optionally, defaults can be applied user-wide by placing a defaults file at `$COLCON_HOME/defaults.yaml` (which is `~/.colcon/defaults.yaml` by default).

In order to manage per-workspace settings, a tool like [`direnv`](https://direnv.net/) can be used to automate the application of the environment variable.

> Once installed and configured with your shell of choice, do the following:

Reviewer: ```suggestion
Once `direnv` is installed and configured with your shell of choice, do the following:
```

Author: fb94592

As an Ignition collection is composed of multiple packages that are frequently built and tested together, `colcon` eases this workflow.

The basic outline of obtaining Ignition source packages via `vcs` and building with `colcon` is available in the [Ignition source installation documentation](https://ignitionrobotics.org/docs/latest/install_ubuntu_src).

### Passing CMake flags via command line

When performing `colcon` builds, flags may be passed to Ignition CMake to configure the build 

This can be done via the `--cmake-args` flag in `colcon`:

```
colcon build --cmake-args -DCMAKE_EXPORT_COMPILE_COMMANDS=1
```

### Passing colcon mixins via command line

To ease configuration of common flags, `colcon` has a concept of `mixins`, that are flags that "shortcut" groups of behavior.

The set of readily-available defaults is in the [colcon-mixin-repository](https://github.com/colcon/colcon-mixin-repository).

To install:

``` 
$ colcon mixin add default https://raw.githubusercontent.com/colcon/colcon-mixin-repository/master/index.yaml
$ colcon mixin update default
$ colcon mixin show
```

To use:

```
colcon build --mixin ccache rel-with-deb-info
```

> 
> 
> 

Reviewer 3: ```suggestion

```

### Using a defaults file

It is useful to be able to apply a consistent set of flags across an entire Ignition collection when building.
One mechanism for accomplishing this is a `defaults.yaml` file.
This is a file of configuration options that `colcon` will read to customize the default behavior.

> More information about the `defaults.yaml` file can be found in the corresponding [`colcon` documentaion](https://colcon.readthedocs.io/en/released/user/configuration.html#defaults-yaml)

Reviewer 3: ```suggestion
More information about the `defaults.yaml` file can be found in the corresponding [`colcon` documentation](https://colcon.readthedocs.io/en/released/user/configuration.html#defaults-yaml)
```
